cpu/stm32f4/periph/adc.c

```diff
@@ -22,164 +22,63 @@
-    adc->CR2 |= ADC_CR2_ADON;
-
-    return 0;
+    return adc_config[line].dev;
 }
 
-int adc_sample(adc_t dev, int channel)
+static inline void poweron(adc_t line)
 {
-    ADC_TypeDef *adc = 0;
-
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            adc = ADC_0_DEV;
-            switch (channel) {
-                case 0:
-                    adc->SQR3 = ADC_0_CH0 & 0x1f;
-                    break;
-                case 1:
-                    adc->SQR3 = ADC_0_CH1 & 0x1f;
-                    break;
-                default:
-                    return -1;
-            }
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            adc = ADC_1_DEV;
-            switch (channel) {
-                case 0:
-                    adc->SQR3 = ADC_1_CH0 & 0x1f;
-                    break;
-                case 1:
-                    adc->SQR3 = ADC_1_CH1 & 0x1f;
-                    break;
-                default:
-                    return -1;
-            }
-            break;
-#endif
-    }
-
-    /* start single conversion */
-    adc->CR2 |= ADC_CR2_SWSTART;
-    /* wait until conversion is complete */
-    while (!(adc->SR & ADC_SR_EOC));
-    /* read and return result */
-    return (int)adc->DR;
+    RCC->APB2ENR |= (1 << adc_config[line].rcc);
 }
 
-void adc_poweron(adc_t dev)
+static inline void poweroff(adc_t line)
 {
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            ADC_0_CLKEN();
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            ADC_1_CLKEN();
-            break;
-#endif
-    }
+    RCC->APB2ENR &= ~(1 << adc_config[line].rcc);
 }
 
-void adc_poweroff(adc_t dev)
+int adc_init(adc_t line)
 {
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            ADC_0_CLKDIS();
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            ADC_1_CLKDIS();
-            break;
-#endif
+    /* check if the linenel is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
     }
+    /* configure the pin */
+    gpio_init_adc(adc_config[line].pin);
+    /* power on the device for the duration of configuration */
+    poweron(line);
+    /* set sequence length to 1 conversion and enable the ADC device */
+    dev(line)->SQR1 = 0;
+    dev(line)->CR2 = ADC_CR2_ADON;
+    /* set clock prescaler */
+    ADC->CCR = (3 << 16);       /* ADC clock = 10,5MHz */
+    /* power off the device again */
+    poweroff(line);
+    return 0;
 }
 
-int adc_map(adc_t dev, int value, int min, int max)
+int adc_sample(adc_t line, adc_res_t res)
 {
-    return (int)adc_mapf(dev, value, (float)min, (float)max);
-}
+    int sample;
 
-float adc_mapf(adc_t dev, int value, float min, float max)
-{
-    return ((max - min) / ((float)adc_config[dev].max_value)) * value;
+    /* check if resolution is applicable */
+    if (res == ADC_RES_14BIT || res == ADC_RES_16BIT) {
+        return -2;
+    }
+    /* power device on for sampling */
+    poweron(line);
```

Comment:
Race condition, if the 2nd thread is waiting in line 71 and doesn't call sleep, the 1ste thread will receive the value of the 2nd thread. Also the 1st thread could `poweroff` the device which would make the 2nd thread end up in an endless loop or read a bogus value.